File: langs/csharp/tests/VMx.Tests/Helpers/RecordedMessages.cs
using System.Reactive.Linq;
using VMx.Messages;

namespace VMx.Tests.Helpers;

/// <summary>
/// Subscribes to an IObservable&lt;IMessage&gt; and records everything
/// observed. Test code asserts against <see cref="Items"/>.
/// </summary>
public sealed class RecordedMessages<TMessage> : IDisposable where TMessage : IMessage
{
    private readonly IDisposable _subscription;

    public List<TMessage> Items { get; } = new();

    public RecordedMessages(IObservable<IMessage> source)
    {
        _subscription = source.OfType<TMessage>().Subscribe(Items.Add);
    }

    public void Dispose() => _subscription.Dispose();
}

File: langs/python/tests/unit/helpers/recorded_messages.py
"""Test helper: subscribe to an Observable[Message] and record everything."""

from __future__ import annotations

from typing import Generic, TypeVar

import reactivex as rx
import reactivex.operators as ops

from vmx.messages.protocols import Message

TMessage = TypeVar("TMessage", bound=Message)


class RecordedMessages(Generic[TMessage]):
    """Wraps an Observable[Message] subscription. Test code asserts on ``.items``."""

    def __init__(self, source: rx.Observable[Message], message_type: type[TMessage]) -> None:
        self.items: list[TMessage] = []
        self._subscription = source.pipe(
            ops.filter(lambda m: isinstance(m, message_type)),
        ).subscribe(self.items.append)

    def dispose(self) -> None:
        self._subscription.dispose()

    def __enter__(self) -> RecordedMessages[TMessage]:
        return self

    def __exit__(self, *exc: object) -> None:
        self.dispose()
